server/src/main/java/org/opensearch/search/profile/query/ConcurrentQueryProfiler.java

```diff
@@ -0,0 +1,124 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.search.profile.query;
+
+import org.apache.lucene.search.Query;
+import org.opensearch.search.profile.ContextualProfileBreakdown;
+import org.opensearch.search.profile.ProfileResult;
+import org.opensearch.search.profile.Timer;
+
+import java.util.ArrayList;
+import java.util.Comparator;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+
+/**
+ * This class acts as a thread-local storage for profiling a query with concurrent execution
+ *
+ * @opensearch.internal
+ */
+public final class ConcurrentQueryProfiler extends QueryProfiler {
+
+    private final Map<Long, ConcurrentQueryProfileTree> threadToProfileTree;
+    private final Map<Long, LinkedList<Timer>> threadToRewriteTimers;
+
+    public ConcurrentQueryProfiler() {
+        super(new ConcurrentQueryProfileTree());
+        long threadId = Thread.currentThread().getId();
+        threadToProfileTree = new ConcurrentHashMap<>();
+        threadToProfileTree.put(threadId, (ConcurrentQueryProfileTree) profileTree);
+        threadToRewriteTimers = new ConcurrentHashMap<>();
+        threadToRewriteTimers.put(threadId, new LinkedList<>());
+    }
+
+    @Override
+    public ContextualProfileBreakdown<QueryTimingType> getQueryBreakdown(Query query) {
+        long threadId = Thread.currentThread().getId();
+        ConcurrentQueryProfileTree profileTree = threadToProfileTree.computeIfAbsent(threadId, k -> new ConcurrentQueryProfileTree());
+        return profileTree.getProfileBreakdown(query);
+    }
+
+    /**
+     * Removes the last (e.g. most recent) element on ConcurrentQueryProfileTree stack.
+     */
+    @Override
+    public void pollLastElement() {
+        long threadId = Thread.currentThread().getId();
+        ConcurrentQueryProfileTree concurrentProfileTree = threadToProfileTree.get(threadId);
+        concurrentProfileTree.pollLast();
+    }
+
+    /**
+     * @return a hierarchical representation of the profiled tree
+     */
+    @Override
+    public List<ProfileResult> getTree() {
+        List<ProfileResult> profileResults = new ArrayList<>();
+        for (Map.Entry<Long, ConcurrentQueryProfileTree> profile : threadToProfileTree.entrySet()) {
+            profileResults.addAll(profile.getValue().getTree());
+        }
+        return profileResults;
+    }
+
+    /**
+     * Begin timing the rewrite phase of a request
+     */
+    @Override
+    public void startRewriteTime() {
+        long threadId = Thread.currentThread().getId();
```

Comment:
Minor, I see that in many places there is the same pattern:

```
long threadId = Thread.currentThread().getId();
...
threadToRewriteTimers.computeIfAbsent(threadId, k -> new LinkedList<>()).add(rewriteTimer);
```

We could simplify it a bit by having a method for the current thread ID as a key:

```
private Long getKey() {
      return Thread.currentThread().getId();
}
```

Than we could just use this method like that:

```
threadToRewriteTimers.computeIfAbsent(getKey(), k -> new LinkedList<>()).add(rewriteTimer);
```